examples/research_projects/robust-speech-event/run_speech_recognition_ctc_streaming.py

```diff
@@ -0,0 +1,645 @@
+    # one local process can concurrently download model & vocab.
+
+    # load feature_extractor and tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(
+        tokenizer_name_or_path,
+        use_auth_token=data_args.use_auth_token,
+        **tokenizer_kwargs,
+    )
+
+    # adapt config
+    config.update(
+        {
+            "feat_proj_dropout": model_args.feat_proj_dropout,
+            "attention_dropout": model_args.attention_dropout,
+            "hidden_dropout": model_args.hidden_dropout,
+            "final_dropout": model_args.final_dropout,
+            "mask_time_prob": model_args.mask_time_prob,
+            "mask_time_length": model_args.mask_time_length,
+            "mask_feature_prob": model_args.mask_feature_prob,
+            "mask_feature_length": model_args.mask_feature_length,
+            "gradient_checkpointing": training_args.gradient_checkpointing,
+            "layerdrop": model_args.layerdrop,
+            "ctc_loss_reduction": model_args.ctc_loss_reduction,
+            "pad_token_id": tokenizer.pad_token_id,
+            "vocab_size": len(tokenizer),
+            "activation_dropout": model_args.activation_dropout,
+        }
+    )
+
+    # create model
+    model = AutoModelForCTC.from_pretrained(
+        model_args.model_name_or_path,
+        cache_dir=model_args.cache_dir,
+        config=config,
+        use_auth_token=data_args.use_auth_token,
+    )
+
+    # freeze encoder
+    if model_args.freeze_feature_encoder:
+        model.freeze_feature_encoder()
+
+    # 6. Now we preprocess the datasets including loading the audio, resampling and normalization
+    audio_column_name = data_args.audio_column_name
+
+    # `phoneme_language` is only relevant if the model is fine-tuned on phoneme classification
+    phoneme_language = data_args.phoneme_language
+
+    # Preprocessing the datasets.
+    # We need to read the audio files as arrays and tokenize the targets.
+    def prepare_dataset(batch):
+        # load audio
+        sample = batch[audio_column_name]
+
+        inputs = feature_extractor(sample["array"], sampling_rate=sample["sampling_rate"])
+        batch["input_values"] = inputs.input_values[0]
+        batch["input_length"] = len(batch["input_values"])
+
+        # encode targets
+        additional_kwargs = {}
+        if phoneme_language is not None:
+            additional_kwargs["phonemizer_lang"] = phoneme_language
+
+        batch["labels"] = tokenizer(batch["target_text"], **additional_kwargs).input_ids
+        return batch
+
+    vectorized_datasets = IterableDatasetDict()
+    with training_args.main_process_first(desc="dataset map preprocessing"):
+        for split, dataset in raw_datasets.items():
+            vectorized_datasets[split] = (
+                dataset.map(
+                    prepare_dataset,
+                )
+                .remove_columns(raw_column_names[split] + ["target_text"])
+                .shuffle(buffer_size=10, seed=training_args.seed)
```

Comment:
buffer_size=10 is a bit small IMO no? @lhoestq what do you think?

Reply:
Also nit maybe split the line in 2,3 lines to make it more readable

Reply:
The buffer_size is a tradeoff. If it's lower then you have a lower shuffling quality, if it's greater the shuffling is better but at the cost of memory and it takes longer for the dataset to start iterating.

For audio data I would say `buffer_size=100` is maybe more reasonable ? Feel free to try it though and see if it's ok in terms of RAM and starting delay.